diff --git a/openssl/crypto/objects/obj_dat.h b/openssl/crypto/objects/obj_dat.h
index d6ac8a446b..9edcbdf2b3 100644
--- a/openssl/crypto/objects/obj_dat.h
+++ b/openssl/crypto/objects/obj_dat.h
@@ -10,7 +10,7 @@
  */
 
 /* Serialized OID's */
-static const unsigned char so[9582] = {
+static const unsigned char so[9600] = {
     0x2A,0x86,0x48,0x86,0xF7,0x0D,                 /* [    0] OBJ_rsadsi */
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,            /* [    6] OBJ_pkcs */
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x02,0x02,       /* [   13] OBJ_md2 */
@@ -1354,9 +1354,11 @@ static const unsigned char so[9582] = {
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,0x09,0x10,0x0D,  /* [ 9549] OBJ_id_smime_ori */
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,0x09,0x10,0x0D,0x03,  /* [ 9559] OBJ_id_smime_ori_kem */
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,0x09,0x10,0x03,0x11,  /* [ 9570] OBJ_id_alg_hss_lms_hashsig */
+    0x2A,0x85,0x03,0x07,0x01,0x01,0x05,0x01,0x03,  /* [ 9581] OBJ_magma_mgm */
+    0x2A,0x85,0x03,0x07,0x01,0x01,0x05,0x02,0x03,  /* [ 9590] OBJ_kuznyechik_mgm */
 };
 
-#define NUM_NID 1502
+#define NUM_NID 1504
 static const ASN1_OBJECT nid_objs[NUM_NID] = {
     {"UNDEF", "undefined", NID_undef},
     {"rsadsi", "RSA Data Security, Inc.", NID_rsadsi, 6, &so[0]},
@@ -2860,9 +2862,11 @@ static const ASN1_OBJECT nid_objs[NUM_NID] = {
     {"id-smime-ori", "id-smime-ori", NID_id_smime_ori, 10, &so[9549]},
     {"id-smime-ori-kem", "id-smime-ori-kem", NID_id_smime_ori_kem, 11, &so[9559]},
     {"id-alg-hss-lms-hashsig", "id-alg-hss-lms-hashsig", NID_id_alg_hss_lms_hashsig, 11, &so[9570]},
+    {"magma-mgm", "magma-mgm", NID_magma_mgm, 9, &so[9581]},
+    {"kuznyechik-mgm", "kuznyechik-mgm", NID_kuznyechik_mgm, 9, &so[9590]},
 };
 
-#define NUM_SN 1493
+#define NUM_SN 1495
 static const unsigned int sn_objs[NUM_SN] = {
      364,    /* "AD_DVCS" */
      419,    /* "AES-128-CBC" */
@@ -3826,6 +3830,7 @@ static const unsigned int sn_objs[NUM_SN] = {
     1012,    /* "kuznyechik-ecb" */
     1183,    /* "kuznyechik-kexp15" */
     1017,    /* "kuznyechik-mac" */
+    1503,    /* "kuznyechik-mgm" */
     1014,    /* "kuznyechik-ofb" */
      477,    /* "lastModifiedBy" */
      476,    /* "lastModifiedTime" */
@@ -3839,6 +3844,7 @@ static const unsigned int sn_objs[NUM_SN] = {
     1187,    /* "magma-ecb" */
     1181,    /* "magma-kexp15" */
     1192,    /* "magma-mac" */
+    1502,    /* "magma-mgm" */
     1189,    /* "magma-ofb" */
      460,    /* "mail" */
      493,    /* "mailPreferenceOption" */
@@ -4359,7 +4365,7 @@ static const unsigned int sn_objs[NUM_SN] = {
     1289,    /* "zstd" */
 };
 
-#define NUM_LN 1493
+#define NUM_LN 1495
 static const unsigned int ln_objs[NUM_LN] = {
      363,    /* "AD Time Stamping" */
      405,    /* "ANSI X9.62" */
@@ -5439,6 +5445,7 @@ static const unsigned int ln_objs[NUM_LN] = {
     1012,    /* "kuznyechik-ecb" */
     1183,    /* "kuznyechik-kexp15" */
     1017,    /* "kuznyechik-mac" */
+    1503,    /* "kuznyechik-mgm" */
     1014,    /* "kuznyechik-ofb" */
     1063,    /* "kx-any" */
     1039,    /* "kx-dhe" */
@@ -5464,6 +5471,7 @@ static const unsigned int ln_objs[NUM_LN] = {
     1187,    /* "magma-ecb" */
     1181,    /* "magma-kexp15" */
     1192,    /* "magma-mac" */
+    1502,    /* "magma-mgm" */
     1189,    /* "magma-ofb" */
      493,    /* "mailPreferenceOption" */
      467,    /* "manager" */
@@ -5856,7 +5864,7 @@ static const unsigned int ln_objs[NUM_LN] = {
      125,    /* "zlib compression" */
 };
 
-#define NUM_OBJ 1350
+#define NUM_OBJ 1352
 static const unsigned int obj_objs[NUM_OBJ] = {
        0,    /* OBJ_undef                        0 */
      181,    /* OBJ_iso                          1 */
@@ -6773,8 +6781,10 @@ static const unsigned int obj_objs[NUM_OBJ] = {
     1122,    /* OBJ_aria_256_ccm                 1 2 410 200046 1 1 39 */
     1174,    /* OBJ_magma_ctr_acpkm              1 2 643 7 1 1 5 1 1 */
     1175,    /* OBJ_magma_ctr_acpkm_omac         1 2 643 7 1 1 5 1 2 */
+    1502,    /* OBJ_magma_mgm                    1 2 643 7 1 1 5 1 3 */
     1177,    /* OBJ_kuznyechik_ctr_acpkm         1 2 643 7 1 1 5 2 1 */
     1178,    /* OBJ_kuznyechik_ctr_acpkm_omac    1 2 643 7 1 1 5 2 2 */
+    1503,    /* OBJ_kuznyechik_mgm               1 2 643 7 1 1 5 2 3 */
     1181,    /* OBJ_magma_kexp15                 1 2 643 7 1 1 7 1 1 */
     1183,    /* OBJ_kuznyechik_kexp15            1 2 643 7 1 1 7 2 1 */
     1148,    /* OBJ_id_tc26_gost_3410_2012_256_paramSetA 1 2 643 7 1 2 1 1 1 */
diff --git a/openssl/crypto/objects/obj_mac.num b/openssl/crypto/objects/obj_mac.num
index e72170b47f..80f1be8eed 100644
--- a/openssl/crypto/objects/obj_mac.num
+++ b/openssl/crypto/objects/obj_mac.num
@@ -1499,3 +1499,5 @@ HKDF_SHA512		1498
 id_smime_ori		1499
 id_smime_ori_kem		1500
 id_alg_hss_lms_hashsig		1501
+magma_mgm		1502
+kuznyechik_mgm		1503
diff --git a/openssl/crypto/objects/objects.txt b/openssl/crypto/objects/objects.txt
index 946cdf5ec6..333981e334 100644
--- a/openssl/crypto/objects/objects.txt
+++ b/openssl/crypto/objects/objects.txt
@@ -1529,9 +1529,11 @@ id-tc26-algorithms 5	: id-tc26-cipher
 id-tc26-cipher 1	:  id-tc26-cipher-gostr3412-2015-magma
 id-tc26-cipher-gostr3412-2015-magma 1	: magma-ctr-acpkm
 id-tc26-cipher-gostr3412-2015-magma 2	: magma-ctr-acpkm-omac
+id-tc26-cipher-gostr3412-2015-magma 3	: magma-mgm
 id-tc26-cipher 2	:  id-tc26-cipher-gostr3412-2015-kuznyechik
 id-tc26-cipher-gostr3412-2015-kuznyechik 1	: kuznyechik-ctr-acpkm
 id-tc26-cipher-gostr3412-2015-kuznyechik 2	: kuznyechik-ctr-acpkm-omac
+id-tc26-cipher-gostr3412-2015-kuznyechik 3	: kuznyechik-mgm
 
 id-tc26-algorithms 6	: id-tc26-agreement
 id-tc26-agreement 1	: id-tc26-agreement-gost-3410-2012-256
diff --git a/openssl/fuzz/oids.txt b/openssl/fuzz/oids.txt
index 5fac3e1c88..886e5cb149 100644
--- a/openssl/fuzz/oids.txt
+++ b/openssl/fuzz/oids.txt
@@ -1350,3 +1350,5 @@ OBJ_HKDF_SHA512="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x03\x1E"
 OBJ_id_smime_ori="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x0D"
 OBJ_id_smime_ori_kem="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x0D\x03"
 OBJ_id_alg_hss_lms_hashsig="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x03\x11"
+OBJ_magma_mgm="\x2A\x85\x03\x07\x01\x01\x05\x01\x03"
+OBJ_kuznyechik_mgm="\x2A\x85\x03\x07\x01\x01\x05\x02\x03"
diff --git a/openssl/include/openssl/evp.h b/openssl/include/openssl/evp.h
index 2fefb1bd93..c20e444481 100644
--- a/openssl/include/openssl/evp.h
+++ b/openssl/include/openssl/evp.h
@@ -386,6 +386,10 @@ typedef struct {
 /* Length of CCM8 tag for TLS */
 #define EVP_CCM8_TLS_TAG_LEN 8
 
+/* GOST TLS 1.3 tag lengths */
+#define EVP_MAGMA_TLS_TAG_LEN 8
+#define EVP_KUZNYECHIK_TLS_TAG_LEN 16
+
 /* Length of tag for TLS */
 #define EVP_CHACHAPOLY_TLS_TAG_LEN 16
 
diff --git a/openssl/include/openssl/obj_mac.h b/openssl/include/openssl/obj_mac.h
index fbaafbaef0..c1f9287d03 100644
--- a/openssl/include/openssl/obj_mac.h
+++ b/openssl/include/openssl/obj_mac.h
@@ -4857,6 +4857,10 @@
 #define NID_magma_ctr_acpkm_omac                1175
 #define OBJ_magma_ctr_acpkm_omac                OBJ_id_tc26_cipher_gostr3412_2015_magma,2L
 
+#define SN_magma_mgm            "magma-mgm"
+#define NID_magma_mgm           1502
+#define OBJ_magma_mgm           OBJ_id_tc26_cipher_gostr3412_2015_magma,3L
+
 #define SN_id_tc26_cipher_gostr3412_2015_kuznyechik             "id-tc26-cipher-gostr3412-2015-kuznyechik"
 #define NID_id_tc26_cipher_gostr3412_2015_kuznyechik            1176
 #define OBJ_id_tc26_cipher_gostr3412_2015_kuznyechik            OBJ_id_tc26_cipher,2L
@@ -4869,6 +4873,10 @@
 #define NID_kuznyechik_ctr_acpkm_omac           1178
 #define OBJ_kuznyechik_ctr_acpkm_omac           OBJ_id_tc26_cipher_gostr3412_2015_kuznyechik,2L
 
+#define SN_kuznyechik_mgm               "kuznyechik-mgm"
+#define NID_kuznyechik_mgm              1503
+#define OBJ_kuznyechik_mgm              OBJ_id_tc26_cipher_gostr3412_2015_kuznyechik,3L
+
 #define SN_id_tc26_agreement            "id-tc26-agreement"
 #define NID_id_tc26_agreement           991
 #define OBJ_id_tc26_agreement           OBJ_id_tc26_algorithms,6L
diff --git a/openssl/include/openssl/tls1.h b/openssl/include/openssl/tls1.h
index e407a55106..af3a50b34b 100644
--- a/openssl/include/openssl/tls1.h
+++ b/openssl/include/openssl/tls1.h
@@ -678,6 +678,12 @@ int SSL_CTX_set_tlsext_ticket_key_evp_cb(SSL_CTX *ctx, int (*fp)(SSL *, unsigned
 #define TLS1_3_CK_SM4_GCM_SM3 0x030000C6
 #define TLS1_3_CK_SM4_CCM_SM3 0x030000C7
 
+/* TLS1.3 GOST ciphersuites from RFC9367 */
+#define TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L 0x0300C103
+#define TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_L 0x0300C104
+#define TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S 0x0300C105
+#define TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_S 0x0300C106
+
 /* a bundle of RFC standard cipher names, generated from ssl3_ciphers[] */
 #define TLS1_RFC_RSA_WITH_AES_128_SHA "TLS_RSA_WITH_AES_128_CBC_SHA"
 #define TLS1_RFC_DHE_DSS_WITH_AES_128_SHA "TLS_DHE_DSS_WITH_AES_128_CBC_SHA"
@@ -872,6 +878,10 @@ int SSL_CTX_set_tlsext_ticket_key_evp_cb(SSL_CTX *ctx, int (*fp)(SSL *, unsigned
 #define TLS1_RFC_RSA_PSK_WITH_ARIA_256_GCM_SHA384 "TLS_RSA_PSK_WITH_ARIA_256_GCM_SHA384"
 #define TLS1_3_RFC_SM4_GCM_SM3 "TLS_SM4_GCM_SM3"
 #define TLS1_3_RFC_SM4_CCM_SM3 "TLS_SM4_CCM_SM3"
+#define TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L"
+#define TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S"
+#define TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L "TLS_GOSTR341112_256_WITH_MAGMA_MGM_L"
+#define TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S "TLS_GOSTR341112_256_WITH_MAGMA_MGM_S"
 
 /*
  * XXX Backward compatibility alert: Older versions of OpenSSL gave some DHE
diff --git a/openssl/ssl/record/methods/recmethod_local.h b/openssl/ssl/record/methods/recmethod_local.h
index d2bb7b394d..77037eaa36 100644
--- a/openssl/ssl/record/methods/recmethod_local.h
+++ b/openssl/ssl/record/methods/recmethod_local.h
@@ -331,7 +331,7 @@ struct ossl_record_layer_st {
 
     /* Flags for GOST ciphers */
     int stream_mac;
-    int tlstree;
+    uint32_t tlstree;
 
     /* TLSv1.3 fields */
     unsigned char *iv; /* static IV */
diff --git a/openssl/ssl/record/methods/tls13_meth.c b/openssl/ssl/record/methods/tls13_meth.c
index ade5739622..d5ce90d500 100644
--- a/openssl/ssl/record/methods/tls13_meth.c
+++ b/openssl/ssl/record/methods/tls13_meth.c
@@ -80,6 +80,26 @@ static int tls13_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
         ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
         return OSSL_RECORD_RETURN_FATAL;
     }
+
+    if (!rl->isdtls && rl->tlstree) {
+        const char *tlstree_mode = NULL;
+
+        if ((rl->tlstree & TLS1_TLSTREE_S) != 0)
+            tlstree_mode = "strong";
+        else if ((rl->tlstree & TLS1_TLSTREE_L) != 0)
+            tlstree_mode = "light";
+
+        if (tlstree_mode != NULL) {
+            params[0] = OSSL_PARAM_construct_utf8_string(
+                OSSL_CIPHER_PARAM_TLSTREE_MODE, (char *)tlstree_mode, 0);
+            params[1] = OSSL_PARAM_construct_end();
+
+            if (!EVP_CIPHER_CTX_set_params(ciph_ctx, params)) {
+                ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+                return OSSL_RECORD_RETURN_FATAL;
+            }
+        }
+    }
 end:
     return OSSL_RECORD_RETURN_SUCCESS;
 }
@@ -164,6 +184,18 @@ static int tls13_cipher(OSSL_RECORD_LAYER *rl, TLS_RL_RECORD *recs,
     for (loop = 0; loop < SEQ_NUM_SIZE; loop++)
         nonce[offset + loop] = staticiv[offset + loop] ^ seq[loop];
 
+    if (!rl->isdtls && rl->tlstree) {
+        OSSL_PARAM params[2];
+
+        params[0] = OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_TLSTREE,
+            seq, 0);
+        params[1] = OSSL_PARAM_construct_end();
+        if (!EVP_CIPHER_CTX_set_params(enc_ctx, params)) {
+            RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+    }
+
     if (!tls_increment_sequence_ctr(rl)) {
         /* RLAYERfatal already called */
         return 0;
diff --git a/openssl/ssl/record/methods/tls_common.c b/openssl/ssl/record/methods/tls_common.c
index 0ffdfe500a..f881902442 100644
--- a/openssl/ssl/record/methods/tls_common.c
+++ b/openssl/ssl/record/methods/tls_common.c
@@ -1256,7 +1256,7 @@ int tls_int_new_record_layer(OSSL_LIB_CTX *libctx, const char *propq, int vers,
             } else if (strcmp(p->key,
                            OSSL_LIBSSL_RECORD_LAYER_PARAM_TLSTREE)
                 == 0) {
-                if (!OSSL_PARAM_get_int(p, &rl->tlstree)) {
+                if (!OSSL_PARAM_get_uint32(p, &rl->tlstree)) {
                     ERR_raise(ERR_LIB_SSL, SSL_R_FAILED_TO_GET_PARAMETER);
                     goto err;
                 }
diff --git a/openssl/ssl/record/rec_layer_s3.c b/openssl/ssl/record/rec_layer_s3.c
index c960aa5893..b247c1833e 100644
--- a/openssl/ssl/record/rec_layer_s3.c
+++ b/openssl/ssl/record/rec_layer_s3.c
@@ -1251,7 +1251,8 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
     BIO *thisbio;
     SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(s);
     const OSSL_RECORD_METHOD *meth;
-    int use_etm, stream_mac = 0, tlstree = 0;
+    int use_etm, stream_mac = 0;
+    uint32_t algorithm2 = 0, tlstree = 0;
     unsigned int maxfrag = (direction == OSSL_RECORD_DIRECTION_WRITE)
         ? ssl_get_max_send_fragment(s)
         : SSL3_RT_MAX_PLAIN_LENGTH;
@@ -1339,6 +1340,14 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
             tlstree = 1;
     }
 
+    if (s->s3.tmp.new_cipher != NULL)
+        algorithm2 = s->s3.tmp.new_cipher->algorithm2;
+    else if (s->session != NULL && s->session->cipher != NULL)
+        algorithm2 = s->session->cipher->algorithm2;
+    else if (s->psksession != NULL && s->psksession->cipher != NULL)
+        algorithm2 = s->psksession->cipher->algorithm2;
+    tlstree |= (algorithm2 & TLS1_TLSTREE_L) | (algorithm2 & TLS1_TLSTREE_S);
+
     if (use_etm)
         *set++ = OSSL_PARAM_construct_int(OSSL_LIBSSL_RECORD_LAYER_PARAM_USE_ETM,
             &use_etm);
@@ -1348,7 +1357,7 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
             &stream_mac);
 
     if (tlstree)
-        *set++ = OSSL_PARAM_construct_int(OSSL_LIBSSL_RECORD_LAYER_PARAM_TLSTREE,
+        *set++ = OSSL_PARAM_construct_uint32(OSSL_LIBSSL_RECORD_LAYER_PARAM_TLSTREE,
             &tlstree);
 
     /*
diff --git a/openssl/ssl/s3_lib.c b/openssl/ssl/s3_lib.c
index e30e56dfeb..8bbc494c93 100644
--- a/openssl/ssl/s3_lib.c
+++ b/openssl/ssl/s3_lib.c
@@ -202,6 +202,80 @@ static SSL_CIPHER tls13_ciphers[] = {
         128,
         128,
     },
+#ifndef OPENSSL_NO_GOST
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_KUZNYECHIK_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_L,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_KUZNYECHIK_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_S,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_MAGMA_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_L,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_MAGMA_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_S,
+        256,
+        256,
+    },
+#endif
 };
 
 /*
diff --git a/openssl/ssl/ssl_ciph.c b/openssl/ssl/ssl_ciph.c
index 7cc99a5c90..c1ff396985 100644
--- a/openssl/ssl/ssl_ciph.c
+++ b/openssl/ssl/ssl_ciph.c
@@ -63,6 +63,8 @@ static const ssl_cipher_table ssl_cipher_table_cipher[SSL_ENC_NUM_IDX] = {
     { SSL_KUZNYECHIK, NID_kuznyechik_ctr_acpkm }, /* SSL_ENC_KUZNYECHIK_IDX 23 */
     { SSL_SM4GCM, NID_sm4_gcm }, /* SSL_ENC_SM4GCM_IDX 24 */
     { SSL_SM4CCM, NID_sm4_ccm }, /* SSL_ENC_SM4CCM_IDX 25 */
+    { SSL_MAGMA_MGM, NID_magma_mgm }, /* SSL_ENC_MAGMA_MGM_IDX 26 */
+    { SSL_KUZNYECHIK_MGM, NID_kuznyechik_mgm }, /* SSL_ENC_KUZNYECHIK_MGM_IDX 27 */
 };
 
 /* NB: make sure indices in this table matches values above */
@@ -311,7 +313,7 @@ int ssl_load_ciphers(SSL_CTX *ctx)
     }
 
 #ifdef OPENSSL_NO_GOST
-    ctx->disabled_enc_mask |= SSL_eGOST2814789CNT | SSL_eGOST2814789CNT12 | SSL_MAGMA | SSL_KUZNYECHIK;
+    ctx->disabled_enc_mask |= SSL_eGOST2814789CNT | SSL_eGOST2814789CNT12 | SSL_MAGMA | SSL_KUZNYECHIK | SSL_MAGMA_MGM | SSL_KUZNYECHIK_MGM;
 #endif
 
     ctx->disabled_mac_mask = 0;
@@ -1801,9 +1803,15 @@ char *SSL_CIPHER_description(const SSL_CIPHER *cipher, char *buf, int len)
     case SSL_MAGMA:
         enc = "MAGMA";
         break;
+    case SSL_MAGMA_MGM:
+        enc = "MAGMAMGM";
+        break;
     case SSL_KUZNYECHIK:
         enc = "KUZNYECHIK";
         break;
+    case SSL_KUZNYECHIK_MGM:
+        enc = "KUZNYECHIKMGM";
+        break;
     case SSL_CHACHA20POLY1305:
         enc = "CHACHA20/POLY1305(256)";
         break;
diff --git a/openssl/ssl/ssl_local.h b/openssl/ssl/ssl_local.h
index 7edef0bd40..0e234e8259 100644
--- a/openssl/ssl/ssl_local.h
+++ b/openssl/ssl/ssl_local.h
@@ -157,6 +157,8 @@
 #define SSL_KUZNYECHIK 0x00800000U
 #define SSL_SM4GCM 0x01000000U
 #define SSL_SM4CCM 0x02000000U
+#define SSL_MAGMA_MGM 0x04000000U
+#define SSL_KUZNYECHIK_MGM 0x08000000U
 
 #define SSL_AESGCM (SSL_AES128GCM | SSL_AES256GCM)
 #define SSL_AESCCM (SSL_AES128CCM | SSL_AES256CCM | SSL_AES128CCM8 | SSL_AES256CCM8)
@@ -242,6 +244,8 @@
  * (currently this also  goes into algorithm2)
  */
 #define TLS1_TLSTREE 0x20000
+#define TLS1_TLSTREE_S 0x80000
+#define TLS1_TLSTREE_L 0x100000
 
 /* Ciphersuite supported in QUIC */
 #define SSL_QUIC 0x00040000U
@@ -361,7 +365,9 @@
 #define SSL_ENC_KUZNYECHIK_IDX 23
 #define SSL_ENC_SM4GCM_IDX 24
 #define SSL_ENC_SM4CCM_IDX 25
-#define SSL_ENC_NUM_IDX 26
+#define SSL_ENC_MAGMA_MGM_IDX 26
+#define SSL_ENC_KUZNYECHIK_MGM_IDX 27
+#define SSL_ENC_NUM_IDX 28
 
 /*-
  * SSL_kRSA <- RSA_ENC
diff --git a/openssl/ssl/statem/statem_lib.c b/openssl/ssl/statem/statem_lib.c
index 46ba024c91..705403716d 100644
--- a/openssl/ssl/statem/statem_lib.c
+++ b/openssl/ssl/statem/statem_lib.c
@@ -497,10 +497,10 @@ MSG_PROCESS_RETURN tls_process_cert_verify(SSL_CONNECTION *s, PACKET *pkt)
 #ifndef OPENSSL_NO_GOST
     if (!SSL_USE_SIGALGS(s)
         && ((PACKET_remaining(pkt) == 64
-                && (EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2001
-                    || EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2012_256))
+                && (EVP_PKEY_is_a(pkey, SN_id_GostR3410_2001)
+                    || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_256)))
             || (PACKET_remaining(pkt) == 128
-                && EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2012_512))) {
+                && EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_512)))) {
         len = (unsigned int)PACKET_remaining(pkt);
     } else
 #endif
@@ -545,11 +545,9 @@ MSG_PROCESS_RETURN tls_process_cert_verify(SSL_CONNECTION *s, PACKET *pkt)
     }
 #ifndef OPENSSL_NO_GOST
     {
-        const char *typename = EVP_PKEY_get0_type_name(pkey);
-        int pktype = OBJ_txt2nid(typename);
-        if (pktype == NID_id_GostR3410_2001
-            || pktype == NID_id_GostR3410_2012_256
-            || pktype == NID_id_GostR3410_2012_512) {
+        if (EVP_PKEY_is_a(pkey, SN_id_GostR3410_2001)
+            || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_256)
+            || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_512)) {
             if ((gost_data = OPENSSL_malloc(len)) == NULL)
                 goto err;
             BUF_reverse(gost_data, data, len);
@@ -1931,8 +1929,6 @@ static int is_tls13_capable(const SSL_CONNECTION *s)
         switch (i) {
         case SSL_PKEY_DSA_SIGN:
         case SSL_PKEY_GOST01:
-        case SSL_PKEY_GOST12_256:
-        case SSL_PKEY_GOST12_512:
             continue;
         default:
             break;
@@ -1949,6 +1945,15 @@ static int is_tls13_capable(const SSL_CONNECTION *s)
         curve = ssl_get_EC_curve_nid(s->cert->pkeys[SSL_PKEY_ECC].privatekey);
         if (tls_check_sigalg_curve(s, curve))
             return 1;
+
+        /*
+         * TODO: There must be an opportunity to check that the sig algs are
+         * consistent with the keys in case if the sig algs are provider-based.
+         * For example, OSSL_FUNC_keymgmt_get_params could return the list of
+         * sig alg code points supported by the key, and this list could be
+         * checked against the sig algs code points stored in
+         * SSL_CTX.sigalg_lookup_cache
+         */
     }
 
     return 0;
diff --git a/openssl/ssl/t1_lib.c b/openssl/ssl/t1_lib.c
index 08c903117f..c0b2fd30de 100644
--- a/openssl/ssl/t1_lib.c
+++ b/openssl/ssl/t1_lib.c
@@ -2406,7 +2406,12 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
     /* Now complete cache and tls12_sigalgs list with provider sig information */
     cache_idx = OSSL_NELEM(sigalg_lookup_tbl);
     for (i = 0; i < ctx->sigalg_list_len; i++) {
+        size_t idx;
         TLS_SIGALG_INFO si = ctx->sigalg_list[i];
+
+        if (!ssl_cert_lookup_by_nid(OBJ_txt2nid(si.sigalg_name), &idx, ctx))
+            goto err;
+
         cache[cache_idx].name = si.name;
         cache[cache_idx].name12 = si.sigalg_name;
         cache[cache_idx].sigalg = si.code_point;
@@ -2414,7 +2419,7 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
         cache[cache_idx].hash = si.hash_name ? OBJ_txt2nid(si.hash_name) : NID_undef;
         cache[cache_idx].hash_idx = ssl_get_md_idx(cache[cache_idx].hash);
         cache[cache_idx].sig = OBJ_txt2nid(si.sigalg_name);
-        cache[cache_idx].sig_idx = (int)(i + SSL_PKEY_NUM);
+        cache[cache_idx].sig_idx = (int)idx;
         cache[cache_idx].sigandhash = OBJ_txt2nid(si.sigalg_name);
         cache[cache_idx].curve = NID_undef;
         cache[cache_idx].mintls = TLS1_3_VERSION;
@@ -3523,10 +3528,8 @@ static int tls12_sigalg_allowed(const SSL_CONNECTION *s, int op,
     if (ssl_cert_is_disabled(SSL_CONNECTION_GET_CTX(s), lu->sig_idx))
         return 0;
 
-    if (lu->sig == NID_id_GostR3410_2012_256
-        || lu->sig == NID_id_GostR3410_2012_512
-        || lu->sig == NID_id_GostR3410_2001) {
-        /* We never allow GOST sig algs on the server with TLSv1.3 */
+    if (lu->sig == NID_id_GostR3410_2001) {
+        /* We never allow GOST2001 sig algs on the server with TLSv1.3 */
         if (s->server && SSL_CONNECTION_IS_TLS13(s))
             return 0;
         if (!s->server
@@ -3562,6 +3565,18 @@ static int tls12_sigalg_allowed(const SSL_CONNECTION *s, int op,
         }
     }
 
+    /* TLS1.2 GOST sig algs could not be negotiated for the use in TLS1.3 */
+    if ((lu->sig == NID_id_GostR3410_2012_256 || lu->sig == NID_id_GostR3410_2012_512)
+        && (((s->server && SSL_CONNECTION_IS_TLS13(s)))
+            || (!s->server
+                && SSL_CONNECTION_GET_SSL(s)->method->version == TLS_ANY_VERSION
+                && s->s3.tmp.min_ver >= TLS1_3_VERSION))
+        && (strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_256_intrinsic_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_512_intrinsic_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512_name) == 0))
+        return 0;
+
     /* Finally see if security callback allows it */
     secbits = sigalg_security_bits(SSL_CONNECTION_GET_CTX(s), lu);
     sigalgstr[0] = (lu->sigalg >> 8) & 0xff;
diff --git a/openssl/ssl/tls13_enc.c b/openssl/ssl/tls13_enc.c
index ae47f5301c..90b1d08cd0 100644
--- a/openssl/ssl/tls13_enc.c
+++ b/openssl/ssl/tls13_enc.c
@@ -391,26 +391,26 @@ static int derive_secret_key_and_iv(SSL_CONNECTION *s, const EVP_MD *md,
         *ivlen = *taglen = (size_t)mac_mdleni;
         *keylen = s->s3.tmp.new_mac_secret_size;
     } else {
+        uint32_t algenc;
 
         *keylen = EVP_CIPHER_get_key_length(ciph);
 
+        if (s->s3.tmp.new_cipher != NULL) {
+            algenc = s->s3.tmp.new_cipher->algorithm_enc;
+        } else if (s->session->cipher != NULL) {
+            /* We've not selected a cipher yet - we must be doing early data */
+            algenc = s->session->cipher->algorithm_enc;
+        } else if (s->psksession != NULL && s->psksession->cipher != NULL) {
+            /* We must be doing early data with out-of-band PSK */
+            algenc = s->psksession->cipher->algorithm_enc;
+        } else {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EVP_LIB);
+            return 0;
+        }
+
         mode = EVP_CIPHER_get_mode(ciph);
         if (mode == EVP_CIPH_CCM_MODE) {
-            uint32_t algenc;
-
             *ivlen = EVP_CCM_TLS_IV_LEN;
-            if (s->s3.tmp.new_cipher != NULL) {
-                algenc = s->s3.tmp.new_cipher->algorithm_enc;
-            } else if (s->session->cipher != NULL) {
-                /* We've not selected a cipher yet - we must be doing early data */
-                algenc = s->session->cipher->algorithm_enc;
-            } else if (s->psksession != NULL && s->psksession->cipher != NULL) {
-                /* We must be doing early data with out-of-band PSK */
-                algenc = s->psksession->cipher->algorithm_enc;
-            } else {
-                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EVP_LIB);
-                return 0;
-            }
             if (algenc & (SSL_AES128CCM8 | SSL_AES256CCM8))
                 *taglen = EVP_CCM8_TLS_TAG_LEN;
             else
@@ -420,6 +420,12 @@ static int derive_secret_key_and_iv(SSL_CONNECTION *s, const EVP_MD *md,
 
             if (mode == EVP_CIPH_GCM_MODE) {
                 *taglen = EVP_GCM_TLS_TAG_LEN;
+#ifndef OPENSSL_NO_GOST
+            } else if ((algenc & SSL_MAGMA_MGM) != 0) {
+                *taglen = EVP_MAGMA_TLS_TAG_LEN;
+            } else if ((algenc & SSL_KUZNYECHIK_MGM) != 0) {
+                *taglen = EVP_KUZNYECHIK_TLS_TAG_LEN;
+#endif
             } else {
                 /* CHACHA20P-POLY1305 */
                 *taglen = EVP_CHACHAPOLY_TLS_TAG_LEN;
diff --git a/openssl/util/perl/OpenSSL/paramnames.pm b/openssl/util/perl/OpenSSL/paramnames.pm
index 62cceaad73..374fe756a7 100644
--- a/openssl/util/perl/OpenSSL/paramnames.pm
+++ b/openssl/util/perl/OpenSSL/paramnames.pm
@@ -165,6 +165,7 @@ my %params = (
     'OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_ENC_LEN' =>            "tls1multi_enclen",    # size_t
 
     'OSSL_CIPHER_PARAM_TLSTREE' => "tlstree", # octet_string
+    'OSSL_CIPHER_PARAM_TLSTREE_MODE' => "tlstree_mode", # utf8_string
 
 # digest parameters
     'OSSL_DIGEST_PARAM_XOFLEN' =>       "xoflen",       # size_t
